homework3/task1/main.c

```diff
@@ -0,0 +1,146 @@
+#include "../../lib/functions.h"
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <stdbool.h>
+
+enum ErrorCode
+{
+    OK,
+    MEMORY_ALLOCATION_ERROR,
+    INVALID_INPUT
+};
+
+void swap(int* number1, int* number2)
+{
+    int temp = *number1;
+    *number1 = *number2;
+    *number2 = temp;
+}
+
+void insertionSort(int* array, const unsigned int end)
+{
+    for (unsigned int i = 1; i < end + 1; ++i)
+    {
+        for (unsigned int j = i; j > 0 && array[j - 1] > array[j]; --j)
+        {
+            swap(&array[j], &array[j - 1]);
+        }
+    }
+}
+
+void smartQSort(int* array, const unsigned int start, const unsigned int end)
+{
+    if (end - start + 1 < 10)
+    {
+        insertionSort(array, end);
+        return;
+    }
+    if (start < end)
+    {
+        int separator = array[end];
+        int left = start;
+        int right = end;
+        while (left <= right)
+        {
+            for (left; array[left] < separator; ++left);
```

Comment:
for (left; ничего не делает. left уже проинициализирован, поэтому
```suggestion
            for (; array[left] < separator; ++left);
```
Ниже то же самое